### PR TITLE
Add ex-ante fork choice test cases

### PR DESCRIPTION
**Pull request by hwwhww**
Thank @casparschwa @djrtwo for the testing plan: https://notes.ethereum.org/m1upI_XcRpSskeG4UvgNtA?view

Differences:
- I combined Scenario 1 and Scenario 2 since the only difference is Step 4.
- [Adversarial attestations > proposer boost] I just use full attestation instead of exact `proposer_boost + 1` votes.
 
It's still verbose. Will think about refactoring tomorrow.

TODOs:
- [x] `spec_configured_state_test` probably is not thread-safe and made `test_proposer_boost_correct_head` failed in CI. Need to check it.
    - sub-PR: #2755, #2765
- [x] Update test generator
    - Either dump `PROPOSER_SCORE_BOOST` to `meta.yaml` or try to dump the full changed config. I'm inclined to the former now.
- ~~Update test format~~
- ~~[DRY] Clean up the test script.~~
- [x] Add ex-ante "sandwich" cases. 
    - sub-PR: #2766

### CODE REVIEW

File: tests/core/pyspec/eth2spec/test/phase0/fork_choice/test_ex_ante.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+@with_presets([MAINNET], reason="to create larger committee")
+def test_ex_ante_attestations_is_greater_than_proposer_boost_without_boost(spec, state):
+    """
+    Adversarial attestations > proposer boost
+
+    Block A - slot N
+    Block B (parent A) - slot N+1
+    Block C (parent A) - slot N+2
+    Attestation_1 (Block B) - slot N+1 – size > proposer_boost
+    """
```
Comment (hwwhww) [inline tests/core/pyspec/eth2spec/test/phase0/fork_choice/test_ex_ante.py:242]:
```suggestion
    """
    
    # For testing `PROPOSER_SCORE_BOOST = 0` case
    yield 'PROPOSER_SCORE_BOOST', 'meta', 0

```

File: tests/core/pyspec/eth2spec/test/phase0/fork_choice/test_ex_ante.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    # On receiving block A at slot `N`
+    block = build_empty_block(spec, state, slot=state.slot + 1)
+    signed_block_a = state_transition_and_sign_block(spec, state, block)
+    yield from tick_and_add_block(spec, store, signed_block_a, test_steps)
+    assert spec.get_head(store) == signed_block_a.message.hash_tree_root()
+
+
+@with_all_phases
+@spec_state_test
+def test_ex_ante_secnario_1_with_boost(spec, state):
```
Comment (casparschwa) [inline tests/core/pyspec/eth2spec/test/phase0/fork_choice/test_ex_ante.py:37]: typo: `test_ex_ante_scenario_1_with_boost(spec, state):` not `test_ex_ante_secnario_1_with_boost(spec, state):`

**Comment by hwwhww** [comment]
### Discussion: about the test vector release

This PR now sets configuration `PROPOSER_SCORE_BOOST:= 0` in some test cases via `@spec_configured_state_test` decorator. I'd suggest that we release these `PROPOSER_SCORE_BOOST:= 0` test vectors *separately* from the regular test vector releases.

Reasons:
- It is only useful for debugging if clients haven't implemented the proposer score boost feature. These test cases can be removed later.
- It introduces a run-time configuration overridden in clients' test suite. Probably not worth complexity for just this case.

/cc @ralexstokes @casparschwa @djrtwo

**Comment by djrtwo** [comment]
I'm okay not releasing proposer_score_boost 0. But it's good to just write a helper that allows for creating attestations as a function of `proposer_score_boost + 1`. This allows us to easily test the boundaries around proposer score boost

(Which is like our tests when score boost = 0 and there is 1 conflicting attstaiton)

**Comment by hwwhww** [comment]
@djrtwo  good points. I made `test_ex_ante_attestations_is_greater_than_proposer_boost_*` cases be tested with minimum attestation participant count such that `attestation_score > proposer_score` now.

if we can release separately, I won't have to change the test format & do deeper refactoring since half of these cases will be deleted.

**Comment by hwwhww** [comment]
@ralexstokes @djrtwo 

I deleted `PROPOSER_SCORE_BOOST: 0` tests from this PR, and created another PR for `PROPOSER_SCORE_BOOST: 0`: #2767 tests. I suggest that:
1. we only merge this PR to the next release.
2. Generate `PROPOSER_SCORE_BOOST: 0` special test vector tarball files from #2767 separately.

File: tests/core/pyspec/eth2spec/test/phase0/fork_choice/test_ex_ante.py [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+@spec_state_test
+def test_ex_ante_attestations_is_greater_than_proposer_boost_with_boost(spec, state):
+    """
+    Adversarial attestations > proposer boost
+
+    Block A - slot N
+    Block B (parent A) - slot N+1
+    Block C (parent A) - slot N+2
+    Attestation_1 (Block B) - slot N+1 – proposer_boost + 1 participants
+    """
```
Comment (casparschwa) [inline tests/core/pyspec/eth2spec/test/phase0/fork_choice/test_ex_ante.py:127]:
```suggestion
    """
    Adversarial attestations > proposer boost
    Objects: 
        Block A - slot N
        Block B (parent A) - slot N+1
        Block C (parent A) - slot N+2
        Attestation_set_1 (Block B); size `proposer_boost + 1` - slot N+1
    Steps:
        Block A received at N — A is head
        Block C received at N+2 — C is head
        Block B received at N+2 — C is head
        Attestation_1 received at N+2 — B is head
    """
```

**Review by casparschwa — COMMENTED** [review]
Mostly on formatting consistency

**Review by djrtwo — APPROVED** [review]

**Pull request merged**